File: apps/chat/frontend/frontend/src/api/friends.js
import { apiFetch } from './client.js'

export function sendFriendRequest({ targetUserId, message = '', source = 'search' }) {
  const id = Number(targetUserId)
  return apiFetch('/friends/request', {
    method: 'POST',
    json: {
      target_user_id: Number.isFinite(id) ? id : targetUserId,
      message,
      source,
    },
  })
}

export function listFriendRequests({ type = 'received', status = 'pending', page = 1, pageSize = 20 }) {
  const q = new URLSearchParams({
    type,
    status,
    page: String(page),
    page_size: String(pageSize),
  })
  return apiFetch(`/friends/requests?${q}`)
}

export function handleFriendRequest(requestId, action) {
  return apiFetch(`/friends/requests/${requestId}`, {
    method: 'PUT',
    json: { action },
  })
}

export function listFriends({ page = 1, pageSize = 50 }) {
  const q = new URLSearchParams({ page: String(page), page_size: String(pageSize) })
  return apiFetch(`/friends?${q}`)
}

export function deleteFriend(friendUserId) {
  return apiFetch(`/friends/${friendUserId}`, { method: 'DELETE' })
}

export function updateFriendRemark(friendUserId, remark) {
  return apiFetch(`/friends/${friendUserId}/remark`, {
    method: 'PUT',
    json: { remark },
  })
}

export function listFriendGroups() {
  return apiFetch('/friends/groups')
}

export function createFriendGroup(name) {
  return apiFetch('/friends/groups', { method: 'POST', json: { name } })
}

export function updateFriendGroup(groupId, body) {
  return apiFetch(`/friends/groups/${groupId}`, { method: 'PUT', json: body })
}

export function deleteFriendGroup(groupId) {
  return apiFetch(`/friends/groups/${groupId}`, { method: 'DELETE' })
}

export function listBlacklist({ page = 1, pageSize = 50 } = {}) {
  const q = new URLSearchParams({ page: String(page), page_size: String(pageSize) })
  return apiFetch(`/friends/blacklist?${q}`)
}

export function addBlacklist(userId) {
  return apiFetch('/friends/blacklist', {
    method: 'POST',
    json: { user_id: Number(userId) },
  })
}

export function removeBlacklist(userId) {
  return apiFetch(`/friends/blacklist/${userId}`, { method: 'DELETE' })
}

export function listWhitelist({ page = 1, pageSize = 50 } = {}) {
  const q = new URLSearchParams({ page: String(page), page_size: String(pageSize) })
  return apiFetch(`/friends/whitelist?${q}`)
}

export function addWhitelist(userId) {
  return apiFetch('/friends/whitelist', {
    method: 'POST',
    json: { user_id: Number(userId) },
  })
}

export function removeWhitelist(userId) {
  return apiFetch(`/friends/whitelist/${userId}`, { method: 'DELETE' })
}

File: apps/chat/backend/backend/chat/urls/friends.py
# chat/urls/friends.py
from django.urls import path
from chat.views.friends import (
    FriendRequestSendView,
    FriendRequestListView,
    FriendRequestHandleView,
    FriendListView,
    FriendDeleteView,
    FriendGroupListView,
    FriendGroupDetailView,
    FriendRemarkView,
    BlacklistView,
    BlacklistRemoveView,
    WhitelistView,
    WhitelistRemoveView,
)

urlpatterns = [
    # 好友列表
    path('', FriendListView.as_view()),                                    # 3.4 GET

    # 好友申请
    path('request', FriendRequestSendView.as_view()),                      # 3.1 POST
    path('requests', FriendRequestListView.as_view()),                     # 3.2 GET
    path('requests/<int:request_id>', FriendRequestHandleView.as_view()),  # 3.3 PUT

    # 好友删除与备注
    path('<int:friend_user_id>', FriendDeleteView.as_view()),              # 3.5 DELETE
    path('<int:friend_user_id>/remark', FriendRemarkView.as_view()),       # 3.10 PUT

    # 好友分组
    path('groups', FriendGroupListView.as_view()),                         # 3.6 GET, 3.7 POST
    path('groups/<int:group_id>', FriendGroupDetailView.as_view()),        # 3.8 PUT, 3.9 DELETE

    # 黑名单
    path('blacklist', BlacklistView.as_view()),                            # 3.11 POST, 3.13 GET
    path('blacklist/<int:user_id>', BlacklistRemoveView.as_view()),        # 3.12 DELETE

    # 白名单
    path('whitelist', WhitelistView.as_view()),                            # 3.14 POST, 3.16 GET
    path('whitelist/<int:user_id>', WhitelistRemoveView.as_view()),        # 3.15 DELETE
]
